### ISSUE TITLE
Support ready for keycloak jwt authentication?

### ISSUE THREAD

**Issue by cdbattags** [zmartzone/lua-resty-openidc#159]
Hi folks!

So I've been following this tutorial:

https://eclipsesource.com/blogs/2018/01/11/authenticating-reverse-proxy-with-keycloak/

And I have an `access_by_lua` block of:

(EDITED)
```
local opts = {
  discovery = "https://auth-devci.poc.learning.amplify.com/auth/realms/Amplify/.well-known/openid-configuration",

  -- the signature algorithm that you expect has been used;
  -- can be a single string or a table.
  -- You should set this for security reasons in order to
  -- avoid accepting a token claiming to be signed by HMAC
  -- using a public RSA key.
  token_signing_alg_values_expected = { "HS256" },

  -- if you want to accept unsigned tokens (using the
  -- "none" signature algorithm) then set this to true.
  accept_none_alg = false,

  -- if you want to reject tokens signed using an algorithm
  -- not supported by lua-resty-jwt set this to false. If
  -- you leave it unset, the token signature will not be
  -- verified at all.
  accept_unsupported_alg = true
}

-- call bearer_jwt_verify for OAuth 2.0 JWT validation
local res, err = require("resty.openidc").bearer_jwt_verify(opts, nil, "pass")

 if err or not res then
  ngx.status = 403
  ngx.say(err and err or "no access_token provided")
  ngx.exit(ngx.HTTP_FORBIDDEN)
end
```

How could I get this working using cookie jwt validation?

**Comment by cdbattags** [comment]
```
2018/05/21 22:27:52 [error] 5#5: *20 lua entry thread aborted: runtime error: /usr/local/openresty/luajit/share/lua/5.1/resty/openidc.lua:367: attempt to concatenate field 'client_secret' (a nil value)
stack traceback:
coroutine 0:
	/usr/local/openresty/luajit/share/lua/5.1/resty/openidc.lua: in function 'openidc_call_token_endpoint'
	/usr/local/openresty/luajit/share/lua/5.1/resty/openidc.lua:878: in function 'authenticate'
```

**Comment by cdbattags** [comment]
Ok, so got the secret working just fine now curious how I might setup jwt auth per user in keycloak.

https://www.nginx.com/blog/authenticating-api-clients-jwt-nginx-plus/

Something similar to what's outlined here?

**Comment by cdbattags** [comment]
I think I'm so close but no context into what this means `jwt signature verification failed: invalid secret type (must be string or function)`

**Comment by cdbattags** [comment]
Finally got it working with:

```
local opts = {
    discovery = "https://auth-devci.poc.learning.amplify.com/auth/realms/Amplify/.well-known/openid-configuration",

    -- the signature algorithm that you expect has been used;
    -- can be a single string or a table.
    -- You should set this for security reasons in order to
    -- avoid accepting a token claiming to be signed by HMAC
    -- using a public RSA key.
    token_signing_alg_values_expected = { "RS256" },

    -- if you want to accept unsigned tokens (using the
    -- "none" signature algorithm) then set this to true.
    accept_none_alg = false,

    -- if you want to reject tokens signed using an algorithm
    -- not supported by lua-resty-jwt set this to false. If
    -- you leave it unset, the token signature will not be
    -- verified at all.
    accept_unsupported_alg = true
}

-- call bearer_jwt_verify for OAuth 2.0 JWT validation
local res, err = require("resty.openidc").bearer_jwt_verify(opts, nil, "pass")

if err or not res then
    ngx.status = 403
    ngx.say(err and err or "no access_token provided")
    ngx.exit(ngx.HTTP_FORBIDDEN)
end
```

**Comment by zandbelt** [comment]
Looks like your box had openssl-dev 1.1.x installed whereas openssl 1.0.x is required.

**Comment by cdbattags** [comment]
Any plans to support latest openssl in the near future? Is this a quick goal in that I could get a fix/PR up if I spend a few days worth of work on it? More than willing to jump into this.

**Comment by zandbelt** [comment]
that's dependent in lua-resty-jwt's support for it

**Comment by cdbattags** [comment]
https://github.com/SkyLothar/lua-resty-jwt/pull/79 to keep this thread updated

**Comment by miconx** [comment]
does anyone have a solution for this problem (updated package maybe?) 
ich have the problem that my keycloak installation is not working anymore because of this :-(

**Comment by cdbattags** [comment]
refer to my latest comment on SkyLothar/lua-resty-jwt#79

away on vacation now but back tomorrow and I'll fork/republish package under new name for luarocks

**Comment by miconx** [comment]
are  opm packages possible too?

**Comment by gene1wood** [comment]
This dependency change was merged into lua-resty-openidc in #165
